presto-main/src/main/java/com/facebook/presto/operator/aggregation/DecimalAverageAggregation.java

```diff
@@ -184,10 +191,25 @@ public static void outputLongDecimal(DecimalType type, BigIntegerAndLongState st
         }
     }
 
-    private static BigDecimal average(BigIntegerAndLongState state, DecimalType type)
+    @VisibleForTesting
+    public static BigDecimal average(LongDecimalWithOverflowAndLongState state, DecimalType type)
     {
-        BigDecimal sum = new BigDecimal(state.getBigInteger(), type.getScale());
+        BigDecimal sum = new BigDecimal(Decimals.decodeUnscaledValue(state.getLongDecimal()), type.getScale());
         BigDecimal count = BigDecimal.valueOf(state.getLong());
+
+        if (state.getOverflow() != 0) {
+            BigInteger overflowMultiplier = TWO.shiftLeft(UNSCALED_DECIMAL_128_SLICE_LENGTH * 8 - 2);
+            BigInteger overflow = overflowMultiplier.multiply(BigInteger.valueOf(state.getOverflow()));
+            sum = sum.add(new BigDecimal(overflow));
+        }
         return sum.divide(count, type.getScale(), ROUND_HALF_UP);
     }
+
+    private static void accumulateAndUpdateOverflow(Slice unscaledValue, LongDecimalWithOverflowAndLongState state)
+    {
+        initializeIfNeeded(state);
+        Slice sum = state.getLongDecimal();
+        long overflow = UnscaledDecimal128Arithmetic.addReturnOverflow(sum, unscaledValue, sum);
+        state.setOverflow(state.getOverflow() + overflow);
```

Comment:
?

`MAX_VALUE + MAX_VALUE` doesn't return `MAX_VALUE` and `overflow = 1`. It returns:

`(MAX_VALUE + MAX_VALUE) % (2^127)` and overflow = `(MAX_VALUE + MAX_VALUE) / (2^127)`. 

With concrete numbers:

MAX_VALUE + MAX_VALUE = 99999999999999999999999999999999999999 + 99999999999999999999999999999999999999 = 29858816539530768268312696284115894270 (with overflow = 1)

then subtract 
- MIN_VALUE -> returns -70141183460469231731687303715884105729L (with overflow 1)

Look at previous response